norns/engine-study-3/lib/FXBusDemo.sc

```diff
@@ -0,0 +1,169 @@
+// SC Bus exercise 4: polls
+// sending brightness + amplitude analysis to Lua
+
+FXBusDemo {
+
+	var <synths;
+	var <busses;
+	var <g;
+
+	*new {
+		^super.new.init();
+	}
+
+	init {
+		var s = Server.default;
+		synths = Dictionary.new;
+		busses = Dictionary.new;
+
+		Routine {
+			busses[\source] = Bus.audio(s, 1);
+			busses[\main_out] = Bus.audio(s, 2);
+			busses[\reverb_send] = Bus.audio(s, 2);
+			busses[\delay_send] = Bus.audio(s, 2);
+
+			// NEW: add an analysis audio bus:
+			busses[\analysis] = Bus.audio(s, 2);
+			// NEW: define control Busses for our Lua polls
+			busses[\brightness] = Bus.control(s);
+			busses[\amp] = Bus.control(s);
+
+			SynthDef.new(\patch_pan, {
+				Out.ar(\out.kr, Pan2.ar(In.ar(\in.kr), \pan.kr(0), \level.kr(1)));
+			}).send(s);
+
+			SynthDef.new(\patch_main, {
+				var src = In.ar(\in.kr, 2);
+				var fc1 = \fc1.kr(600);
+				var fc2 = \fc2.kr(1800);
+
+				var ampLo = \ampLo.kr(1);
+				var ampMid = \ampMid.kr(1);
+				var ampHi = \ampHi.kr(1);
+
+				var lo = LPF.ar(LPF.ar(src, fc1), fc1) * ampLo;
+				var mid = HPF.ar(HPF.ar(LPF.ar(LPF.ar(src, fc2), fc2), fc1), fc1) * ampMid;
+				var hi = HPF.ar(HPF.ar(src, fc2), fc2) * ampHi;
+
+				var mix = lo + mid + hi;
+
+				Out.ar(\out.kr, mix * \level.kr(1));
+			}).send(s);
+
+			// add a group to order our synths / nodes:
+			g = Group.new(s);
+
+			// define our source synth:
+			synths[\source] = SynthDef.new(\sourceBlip, {
+				var snd = LPF.ar(Saw.ar(\hz.kr(330)), \fchz.kr(800).clip(20,20000));
+				snd = snd * LagUD.ar(Impulse.ar(0.3), 0, 10);
+				Out.ar(\out.kr, snd * \level.kr(0.5));
+			}).play(target:g, addAction:\addToTail, args:[
+				\out, busses[\source]
+			]);
+
+			// why are we syncing here? two reasons:
+			// 1. so the common SynthDefs above are present on the Server when requested
+			// 2. because the send synths below use \addToTail,
```

Comment:
i may have indicated at some time that (2) could be the case, but in fact it isn't: node ID allocation happens entirely on the client side, so e.g. this works fine with no need for pausing:

```
a = Synth.new(...); 
b = Synth.new(..., target:a, addAction:\addBefore); 
c = Synth.new(..., target:a, addAction:\addAfter);
```

resulting in node order `a, b, c`. it should also work with complex hierarchies.

(under the hood: the client uses a simple counter to get a new node ID for each new synth. node IDs are sent as targets along with `doneAction` code values in the OSC bundle instructing the server. those IDs and connecting verbs completely describe the node graph, so there's no need for further explicit sync.) 

so, `s.sync` is really necessary in this example only to wait for the synthdef to be sent. (it can also be used to wait for buffer allocation and probably other things i'm forgetting, which don't apply here.) if you don't want to introduce s.sync here there is probably another way to add the synthdefs that implicitly blocks. it _is_ important to emphasize node ordering at this point.

Reply:
ah gotcha, that ordering note makes sense, thank you!

re: `s.sync`, i just want to confirm my understanding -- in this file, it's needed:

- after `synths[\source]` is `play`'d, so we start our source sound up ahead of all the processing
- after `synths[\reverb]` is `play`'d, so we initialize the audio chain ahead of `synths[\main_out]`

and it's *not* needed between `synths[\main_out]` and `synths[\brightness]` because `synths[\main_out]` is only invoking a known SynthDef (`\patch_main`).

this all *sounds* right when i run the sketch, but is this understanding correct? or are there other places where syncs are necessary?

thank you so so much!!

Reply:
`.play` is confusing. you're absolutely right that the actual creation of the synth is asynchronous and doesn't happen immediately. let's look at the implementation:

```

	play { arg target,args,addAction=\addToHead;
		var synth, msg;
//		this.deprecated(thisMethod, Function.findRespondingMethodFor(\play));
		target = target.asTarget;
		synth = Synth.basicNew(name,target.server);
		msg = synth.newMsg(target, args, addAction);
		this.send(target.server, msg);
		^synth
	}

```

yes, the actual creation of the synth is initiated by a pre-built OSC bundle passed as `completionMsg` to `SynthDef.send`.


but! notice that the client-side `Synth` object returns immediately! that is because in fact, _node ordering is completely determined by the client._ and with `.play`, the client knows everything it needs to know about the new node while both the comppiled SynthDef data and the request for the new node instantiation are still in flight:

as further demonstration, this works:

```
( // silent: (x -> y -> z), so y overwrites with silence
b = Bus.audio(s, 2);
{ b.scope }.defer;

x = SynthDef.new(\brown, { Out.ar(\out.kr, BrownNoise.ar.dup * \amp.kr(0.1)) })
.play(args:[\out, b]);

y = SynthDef.new(\silence_insert, { ReplaceOut.ar(\bus.kr, Silent.ar(2)) })
.play(target: x, addAction:\addAfter, args:[\bus, b]);

z = SynthDef.new(\patch_stereo, { Out.ar(\out.kr(0), In.ar(\in.kr, 2)) })
.play(target: s, addAction:\addToTail, args:[\in, b]);
)

( // not silent: default node ordering puts each new node at the head
b = Bus.audio(s, 2);
{ b.scope }.defer; // still _looks_ silent, because the bus is being overwritten!

x = SynthDef.new(\brown, { Out.ar(\out.kr, BrownNoise.ar.dup * \amp.kr(0.1)) })
.play;

y = SynthDef.new(\silence_insert, { ReplaceOut.ar(\bus.kr, Silent.ar(2)) })
.play;

z = SynthDef.new(\patch_stereo, { Out.ar(\out.kr(0), In.ar(\in.kr, 2)) })
.play(target: s, addAction:\addToTail, args:[\in, b]);
)
```

-- 

so anyways. i just think for the purpose of a study / best practices document like this, it is super important to be clear and accurate about each detail that is communicated. the asynchronous architecture is super confusing at the best of times.

in this case i think we could strip the use of `.sync` down to the mininum (i believe it is only needed for `patch_pan`) and just mention its necessity there, rather than in general terms. (i'm even inclined to restructure it without `.sync` and with a callback instead, which doesn't scale as well but perhaps avoids some darker waters in the simple case.)

Reply:
I was reminded of another need for sync. If you create a synth, and immediately call `Synth.set` on it, the parameter set will (/may?) without `s.sync` first.

(I wonder if sometimes the two commands are handled on consecutive audio blocks, this working, and sometimes on same block, not working.. something like that.)